File: implementations/nestjs/examples/src/account/domain/money.ts
import { AccountErrorMessage as ErrorMessage } from '@/account/account-error-message'

export class Money {
  public readonly amount: number
  public readonly currency: string

  constructor(params: { amount: number; currency: string }) {
    if (params.amount < 0) throw new Error(ErrorMessage['The amount must be 0 or greater.'])
    this.amount = params.amount
    this.currency = params.currency
  }

  public add(other: Money): Money {
    this.assertSameCurrency(other)
    return new Money({ amount: this.amount + other.amount, currency: this.currency })
  }

  public subtract(other: Money): Money {
    this.assertSameCurrency(other)
    return new Money({ amount: this.amount - other.amount, currency: this.currency })
  }

  public isLessThan(other: Money): boolean {
    this.assertSameCurrency(other)
    return this.amount < other.amount
  }

  public isZero(): boolean {
    return this.amount === 0
  }

  public equals(other: Money): boolean {
    return this.amount === other.amount && this.currency === other.currency
  }

  private assertSameCurrency(other: Money): void {
    if (this.currency !== other.currency) throw new Error(ErrorMessage['The currencies do not match.'])
  }
}

File: implementations/go/examples/internal/domain/account/money.go
package account

type Money struct {
	Amount   int64
	Currency string
}

func NewMoney(amount int64, currency string) (Money, error) {
	if amount < 0 {
		return Money{}, ErrInvalidMoneyAmount
	}
	return Money{Amount: amount, Currency: currency}, nil
}

func (m Money) Add(other Money) (Money, error) {
	if m.Currency != other.Currency {
		return Money{}, ErrCurrencyMismatch
	}
	return Money{Amount: m.Amount + other.Amount, Currency: m.Currency}, nil
}

func (m Money) Subtract(other Money) (Money, error) {
	if m.Currency != other.Currency {
		return Money{}, ErrCurrencyMismatch
	}
	return Money{Amount: m.Amount - other.Amount, Currency: m.Currency}, nil
}

func (m Money) LessThan(other Money) bool {
	return m.Amount < other.Amount
}

func (m Money) IsZero() bool {
	return m.Amount == 0
}

func (m Money) Equals(other Money) bool {
	return m.Amount == other.Amount && m.Currency == other.Currency
}
